src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1215,7 +1215,68 @@ module T = struct
       =
     (List.map ~f:(With_status.map ~f:Transaction.forget) a, b, c, d)
 
-  let check_commands ledger ~verifier (cs : User_command.t With_status.t list) =
+  type transaction_pool_proxy =
+    { find_by_hash :
+           Mina_transaction.Transaction_hash.t
+        -> Mina_transaction.Transaction_hash.User_command_with_valid_signature.t
+           option
+    }
+
+  let dummy_transaction_pool_proxy : transaction_pool_proxy =
+    { find_by_hash = const None }
+
+  (* NOTE: when we have access to txn pool, any txns in pool should be already
+     verified, hence we could utilize this fact to perform a faster version of
+     command verification *)
+  let verify_command_with_transaction_pool_proxy
+      (transaction_pool_proxy : transaction_pool_proxy)
+      (cmd_with_status : User_command.Verifiable.t With_status.t) =
+    let With_status.{ data = verifiable_cmd; _ } = cmd_with_status in
+    let cmd_hash =
+      verifiable_cmd |> User_command.of_verifiable
+      |> Transaction_hash.hash_command
+    in
+    match transaction_pool_proxy.find_by_hash cmd_hash with
+    | None ->
+        `No_fast_forward
+    | Some _ ->
+        Verifier.Common.verify_command_from_mempool cmd_with_status
+
+  let process_separately (type input left right output)
+      ~(partitioner : input -> (left, right) Core_kernel.Either.t)
+      ~(process_left : left list -> output list Deferred.Or_error.t)
```

Comment:
Given that constructors are named `First` and `Second`, maybe it makes sense to name the parameter as something like `process_first_partition`. A bit longer, but more informative. WDYT?